devito/finite_differences/finite_difference.py

```diff
@@ -239,59 +247,120 @@ def staggered_diff(f, dim, order, stagger=centered, theta=0, phi=0):
     :param theta: Dip (or polar) angle for rotated FD
     :param phi: Azimuth angle for rotated FD
     """
-    ndim = len(f.space_dimensions)
-    off = dict([(d, 0) for d in f.space_dimensions])
     if stagger == left:
-        off[dim] = -.5
+        off = -.5
     elif stagger == right:
-        off[dim] = .5
+        off = .5
     else:
-        off[dim] = 0
-
-    if theta == 0 and phi == 0:
-        diff = dim.spacing
-        idx = list(set([(dim + int(i+.5+off[dim])*diff)
-                        for i in range(-int(order / 2), int(order / 2))]))
-        if int(order / 2) == 1:
-            idx = [dim - diff, dim]
-        deriv = f.diff(dim).as_finite_difference(idx, x0=dim + off[dim]*dim.spacing)
-        return deriv.evalf(_PRECISION)
-    else:
-        if ndim < 2 or ndim > 3:
-            error("Only 2D and 3D supports rotated finite difference,"
-                  " %d dimensions as input" % ndim)
-            raise ValueError("Non supported number of dimensions")
-
-        x = f.space_dimensions[0]
-        z = f.space_dimensions[-1]
-        idxx = list(set([(x + int(i+.5+off[x])*x.spacing)
-                         for i in range(-int(order / 2), int(order / 2))]))
-        dx = f.diff(x).as_finite_difference(idxx, x0=x + off[x]*x.spacing)
-        dx = dx.evalf(_PRECISION)
-
-        idxz = list(set([(z + int(i+.5+off[z])*z.spacing)
-                         for i in range(-int(order / 2), int(order / 2))]))
-        dz = f.diff(z).as_finite_difference(idxz, x0=z + off[z]*z.spacing)
-        dz = dx.evalf(_PRECISION)
-
-        dy = 0
-        is_y = False
-
-        if ndim == 3:
-            y = f.space_dimensions[1]
-            idxy = list(set([(y + int(i+.5+off[y])*y.spacing)
-                             for i in range(-int(order / 2), int(order / 2))]))
-            dy = f.diff(y).as_finite_difference(idxy, x0=y + off[y]*y.spacing)
-            dy = dy.evalf(_PRECISION)
-            is_y = (dim == y)
-
-        if dim == x:
-            return cos(theta) * cos(phi) * dx + sin(phi) * cos(theta) * dy -\
-                sin(theta) * dz
-        elif dim == z:
-            return sin(theta) * cos(phi) * dx + sin(phi) * sin(theta) * dy +\
-                cos(theta) * dz
-        elif is_y:
-            return -sin(phi) * dx + cos(phi) * dy
+        off = 0
+    diff = dim.spacing
+    idx = list(set([(dim + int(i+.5+off)*diff)
+                    for i in range(-int(fd_order / 2), int(fd_order / 2))]))
+    if int(fd_order / 2) == 1:
+        idx = [dim - diff, dim]
+    deriv = f.diff(*(tuple(dim for _ in range(deriv_order))))
+    return deriv.as_finite_difference(idx, x0=dim + off*dim.spacing).evalf(_PRECISION)
+
+
+def initialize_derivatives(self):
+    """
+    Dynamically create notational shortcuts for space derivatives.
+    """
+    start_space = 0
+    if self.is_TimeFunction:
```

Comment:
let's think about this: what makes ``time`` so special here that we have to do ``if is_TimeFunction``... can we generalise this?

Reply:
OK, after some more thinking... how about we do the following instead. 

We keep ``initialize_derivatives`` a free function (ie, we don't put it in a mixin class). But now we let it accept two parameters, a ``dimension`` and an (approximation) ``order``. Isn't this the only piece of information that we actually need to create the shortcuts? Then, ``initialize_derivatives`` would either return a bag of functions/shortcuts that the caller adds to itself, or we could also pass the object (i.e., a Function) to which we want the newly-created shortcuts to be attached .

If we do that, we don't need to distinguish anymore between time and space, I think, which would be invaluable in terms of code organisation and clarity

Within ``Function/TimeFunction.__init__``, we could then have something along the lines of
```
for d in dimensions:
    initialize_derivatives(d, d_order, self)
```
or, equivalently:
```
for d in dimensions:
    for shortcut in generate_derivative_shortcuts(d, d_order):
         setattr(self, shortcut.__name__, shortcut)
```
or something like that...

thoughts? Am I off the track?